### PR TITLE
[DAG] SimplifyMultipleUseDemandedBits - bypass ADD nodes if either operand is zero

### PR DESCRIPTION
**Pull request by RKSimon**
The dpbusd_const.ll test change is due to us losing the expanded add reduction pattern as one of the elements is known to be zero (removing one of the adds from the reduction pyramid). I don't think its of concern.

Noticed while working on #107423

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/srem.ll [hunk 1]
```diff
@@ -39,8 +39,8 @@ define amdgpu_kernel void @srem_i16_7(ptr addrspace(1) %out, ptr addrspace(1) %i
 ; TAHITI-NEXT:    s_waitcnt vmcnt(0)
 ; TAHITI-NEXT:    v_readfirstlane_b32 s0, v0
 ; TAHITI-NEXT:    s_mulk_i32 s0, 0x4925
-; TAHITI-NEXT:    s_lshr_b32 s1, s0, 31
 ; TAHITI-NEXT:    s_ashr_i32 s0, s0, 17
+; TAHITI-NEXT:    s_bfe_u32 s1, s0, 0x1000f
```
Comment (RKSimon) [inline llvm/test/CodeGen/AMDGPU/srem.ll:43]: @arsenm @jayfoad Have we missed a fold here? Is BFE better than SRL?

File: llvm/test/CodeGen/X86/dpbusd_const.ll [hunk 2]
```diff
@@ -55,7 +55,7 @@ define i32 @mul_4xi8_zc(<4 x i8> %a, i32 %c) {
 ; AVX512VLVNNI-NEXT:    retq
 entry:
   %0 = zext <4 x i8> %a to <4 x i32>
-  %1 = mul nsw <4 x i32> %0, <i32 0, i32 1, i32 2, i32 127>
+  %1 = mul nsw <4 x i32> %0, <i32 16, i32 1, i32 2, i32 127>
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/dpbusd_const.ll:58]: Why changing test case?

Comment (phoebewang) [reply llvm/test/CodeGen/X86/dpbusd_const.ll:58]: Oh, I saw the comment now.

**Comment by RKSimon** [comment]
any more comments?

File: llvm/test/CodeGen/AMDGPU/srem.ll (hunk 1 above)
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/srem.ll:43]: This is worse. The BFE has a bigger encoding because it needs to embed the weird immediate

Comment (RKSimon) [reply llvm/test/CodeGen/AMDGPU/srem.ll:43]: OK - I'll see if I can find the missing fold.

File: llvm/test/CodeGen/RISCV/div-by-constant.ll [hunk 3]
```diff
@@ -755,8 +755,9 @@ define i16 @sdiv16_constant_srai(i16 %a) nounwind {
 ; RV32IM-NEXT:    lui a1, 6
 ; RV32IM-NEXT:    addi a1, a1, 1639
 ; RV32IM-NEXT:    mul a0, a0, a1
-; RV32IM-NEXT:    srli a1, a0, 31
 ; RV32IM-NEXT:    srai a0, a0, 17
+; RV32IM-NEXT:    slli a1, a0, 16
```
Comment (arsenm) [inline llvm/test/CodeGen/RISCV/div-by-constant.ll:759]: this is also 1 -> 2 instructions?

Comment (RKSimon) [reply llvm/test/CodeGen/RISCV/div-by-constant.ll:759]: yes its the same regression

**Review by arsenm — APPROVED** [review]

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 4]
```diff
@@ -797,6 +797,16 @@ SDValue TargetLowering::SimplifyMultipleUseDemandedBits(
       return Op.getOperand(1);
     break;
   }
+  case ISD::ADD: {
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:800]: Should this also cover or and sub?

Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:800]: OR is already handled more generally - I'll see if I can get (SUB X, 0) to match in any tests

**Pull request merged**